pss/crypto/crypto.go

```diff
@@ -0,0 +1,656 @@
+	if err != nil {
+		return nil, err
+	}
+	salt, err := crypto.generateSecureRandomData(aesNonceLength) // never use more than 2^32 random nonces with a given key
+	if err != nil {
+		return nil, err
+	}
+	encrypted := aesgcm.Seal(nil, salt, rawBytes, nil)
+	encBytes := append(encrypted, salt...)
+	return encBytes, nil
+}
+
+// signHash calculates an ECDSA signature.
+func (crypto *defaultCryptoBackend) signHash(hash []byte, prv *ecdsa.PrivateKey) (sig []byte, err error) {
+	return ethCrypto.Sign(hash, prv)
+}
+
+// Obtain public key from the signed message and the signature
+func (crypto *defaultCryptoBackend) sigToPub(signed, sig []byte) (*ecdsa.PublicKey, error) {
+	hash := crypto.keccak256(signed)
+	return ethCrypto.SigToPub(hash, sig)
+}
+
+// GenerateRandomID generates a random string, which is then returned to be used as a key id
+func (crypto *defaultCryptoBackend) generateRandomID() (id string, err error) {
+	buf, err := crypto.generateSecureRandomData(keyIDSize)
+	if err != nil {
+		return "", err
+	}
+	if !validateDataIntegrity(buf, keyIDSize) {
+		return "", fmt.Errorf("error in generateRandomID: crypto/rand failed to generate random data")
+	}
+	id = common.Bytes2Hex(buf)
+	return id, err
+}
+
+// keccak256 calculates and returns the keccak256 hash of the input data.
+func (crypto *defaultCryptoBackend) keccak256(data ...[]byte) []byte {
+	return ethCrypto.Keccak256(data...)
+}
+
+// generateSecureRandomData generates random data where extra security is required.
+// The purpose of this function is to prevent some bugs in software or in hardware
+// from delivering not-very-random data. This is especially useful for AES nonce,
+// where true randomness does not really matter, but it is very important to have
+// a unique nonce for every message.
+func (crypto *defaultCryptoBackend) generateSecureRandomData(length int) ([]byte, error) {
+	x := make([]byte, length)
+	y := make([]byte, length)
+	res := make([]byte, length)
+
+	_, err := crand.Read(x)
+	if err != nil {
+		return nil, err
+	} else if !validateDataIntegrity(x, length) {
+		return nil, errSecureRandomData
+	}
+	_, err = mrand.Read(y)
+	if err != nil {
+		return nil, err
+	} else if !validateDataIntegrity(y, length) {
+		return nil, errSecureRandomData
+	}
+	for i := 0; i < length; i++ {
+		res[i] = x[i] ^ y[i]
+	}
+	if !validateDataIntegrity(res, length) {
+		return nil, errSecureRandomData
+	}
+	return res, nil
+}
+
+func (crypto *defaultCryptoBackend) importECDSAPublic(key *ecdsa.PublicKey) *ecies.PublicKey {
+	return ecies.ImportECDSAPublic(key)
+}
+
+func (crypto *defaultCryptoBackend) GenerateKey() (*ecdsa.PrivateKey, error) {
+	return ethCrypto.GenerateKey()
+}
+
+// CryptoUtils
+
+// NewKeyPair generates a new cryptographic identity for the client, and injects
+// it into the known identities for message decryption. Returns ID of the new key pair.
+func (crypto *defaultCryptoBackend) NewKeyPair(ctx context.Context) (string, error) {
```

Comment:
what is context  doing here?

Reply:
It is an interface thing. In Whisper, depending on the implementation, the context could be used or not. In this case is not used, but I guess some implementation can time-out generating keys.

Reply:
I will remove it anyway.